README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# eltr <img src='man/figures/logo.png' align="right" height="139" />

<!-- badges: start -->

[![Lifecycle:
experimental](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://www.tidyverse.org/lifecycle/#experimental)
[![R build
status](https://github.com/RandhirBilkhu/eltr/workflows/R-CMD-check/badge.svg)](https://github.com/RandhirBilkhu/eltr/actions)
[![Codecov test
coverage](https://codecov.io/gh/RandhirBilkhu/eltr/branch/main/graph/badge.svg)](https://codecov.io/gh/RandhirBilkhu/eltr?branch=main)
<!-- badges: end -->

The goal of eltr is to help analysis of catastrophe model outputs.

## Installation

You can install the development version from
[GitHub](https://github.com/RandhirBilkhu/eltr) with:

``` r
# install.packages("devtools")
devtools::install_github("RandhirBilkhu/eltr")
```

## Example

This is a basic example which shows you how to solve a common problem:

``` r
library(eltr)
## load a rawelt file

raw_elt <- eltr::example_elt

raw_elt
#>     id rate  mean sdevi sdevc     exp
#>  1:  1 0.10   500   500   200  100000
#>  2:  2 0.10   200   400   100    5000
#>  3:  3 0.20   300   200   400   40000
#>  4:  4 0.10   100   300   500    4000
#>  5:  5 0.20   500   100   200    2000
#>  6:  6 0.25   200   200   500   50000
#>  7:  7 0.01  1000   500   600  100000
#>  8:  8 0.12   250   300   100    5000
#>  9:  9 0.14  1000   500   200    6000
#> 10: 10 0.00 10000  1000   500 1000000

## paramterise elt

elt <- create_elt(raw_elt, ann_rate="rate", mu="mean", sdev_i = "sdevi" , sdev_c = "sdevc", expval = "exp")

## generate a YLT 

ylt <- create_ylt(elt, sims=10 ,ann_rate = "rate" , event_id = "id", expval = "exp" , mu ="mean")

ylt
#>     Year         Loss Event
#>  1:    1 0.000000e+00  None
#>  2:    2 2.963659e+02     5
#>  3:    3 4.353146e-09     6
#>  4:    4 3.086963e+02     8
#>  5:    4 3.337206e+02     3
#>  6:    4 1.429257e-03     2
#>  7:    5 0.000000e+00  None
#>  8:    6 5.525234e+00     3
#>  9:    6 4.838321e+00     8
#> 10:    6 8.642469e+02     7
#> 11:    7 4.338468e+02     3
#> 12:    7 1.357392e+03     2
#> 13:    7 1.317510e+02     1
#> 14:    8 1.204985e+03     5
#> 15:    8 4.033199e+02     8
#> 16:    9 6.361959e-02     6
#> 17:   10 0.000000e+00  None

### calculate AAL and OEP

ann <-ylt[ ,Loss :=sum(Loss) , by=Year]

ep <- create_oep_curve(ann , y= "Year", z="Loss")
```

<a target="_blank" href="https://icons8.com/icons/set/hurricane">Hurricane
icon</a> icon by <a target="_blank" href="https://icons8.com">Icons8</a>


NEWS.md
# eltr 0.1.0

* `create_elt` transforms a raw elt into a object with beta distribution parameters for secondary uncertainty

* `create_ylt` runs a monte carlo simulation assuming a poisson distribution for primary uncertainty

* `create_oep_curve` returns the occurrence loss at specified return periods

* `layer_loss` calculates limited loss to layer




tests/testthat.R
library(testthat)
library(eltr)

test_check("eltr")


tests/testthat/test-cat_functions.R
test_that("multiplication works", {
  expect_equal(2 * 2, 4)
})


testthat::test_that("layer loss is correct", {
  expect_equal(layer_loss(10,2,5),5)

})

set.seed(1)
elt <- create_elt(example_elt, ann_rate="rate", mu="mean", sdev_i = "sdevi" , sdev_c = "sdevc", expval = "exp")
ylt <- create_ylt(elt, sims=10 ,ann_rate = "rate" , event_id = "id", expval = "exp" , mu ="mean")
ann <-ylt[ ,Loss :=sum(Loss) , by=Year]
ep <- create_oep_curve(ann , y= "Year", z="Loss")


testthat::test_that("elt normalised rate correct",{
  expect_equal(   sum(elt$random_num) ,1)

})


testthat::test_that("ylt creation",{
  expect_equal(   sum(ylt$Loss),  13833.1, tolerance=0.01)

})


testthat::test_that("OEP",{
  expect_equal(   sum(ep$OEP),  19883.16 , tolerance=0.01)

})



tests/testthat/test_cat_functions.R
context("function testing")

library(catr)
set.seed(1)

elt <- create_elt(example_elt, ann_rate="rate", mu="mean", sdev_i = "sdevi" , sdev_c = "sdevc", expval = "exp")
ylt <- create_ylt(elt, sims=10 ,ann_rate = "rate" , event_id = "id", expval = "exp" , mu ="mean")
ann <-ylt[ ,Loss :=sum(Loss) , by=Year]

ep <- create_ep_curve(ann , y= "Year", z="Loss")

testthat::test_that("layer loss is correct", {
  expect_equal(layer_loss(10,2,5),5)

})


testthat::test_that("elt normalised rate correct",{
  expect_equal(   sum(elt$random_num) ,1)

})


testthat::test_that("ylt creation",{
  expect_equal(   sum(ylt$Loss),  13833.1, tolerance=0.01)

})


testthat::test_that("OEP",{
  expect_equal(   sum(ep$OEP),  167425.6 , tolerance=0.01)

})




R/cat_functions.R
.datatable.aware= TRUE

#' Limited loss to the layer
#'
#' @param x event loss
#' @param Excess treaty retention
#' @param Limit treaty limit
#'
#' @return limited loss to the layer
#' @examples
#' layer_loss(5,2,6)
#' layer_loss(5,10,6)
#'
#' @export

layer_loss <- function(x, Excess, Limit){
  pmin(pmax(x - Excess, 0), Limit)
}



#' Create parameters for ELT simulation
#'
#' @param dt an ELT (Event Loss Table)
#' @param ann_rate a vector of annual rates for each event
#' @param mu a vector of mean event loss
#' @param sdev_i a vector of independent standard deviations
#' @param sdev_c a vector of correlated standard deviations
#' @param expval the total values exposed in each event
#'
#' @return a data.table object with mean damage ratio, total standard deviation and alpha/beta parameters
#' @examples
#' create_elt (eltr::example_elt, ann_rate="rate", mu="mean", 
#'            sdev_i = "sdevi" , sdev_c = "sdevc", expval = "exp")
#' 
#' @export

create_elt <- function(dt, ann_rate, mu , sdev_i, sdev_c, expval) {

  mdr <- sdev <- cov <- alpha <- beta  <- random_num <- NULL # avoid the no visible binding for global variable when running R_CMD_CHECK


  dt[ , mdr := get(mu)/get(expval)]
  dt[, sdev := get(sdev_i) + get(sdev_c)]
  dt[, cov :=  sdev /get(mu) ]
  dt[, alpha := ( (1- mdr)/ (cov^2) - mdr) ]

  dt$alpha[!is.finite(dt$alpha)] <- 0

  dt[ , beta := (alpha * (1- mdr))/mdr  ]

  lda <- sum(dt[, get(ann_rate)] )

  dt[ , random_num := get(ann_rate)/ lda]


  data.table::data.table(dt)

}




#' Create a YLT from ELT via monte carlo simulation
#'
#' @param dt a data.table with modified ELT
#' @param sims number of years to simulate
#' @param ann_rate event frequency
#' @param event_id unique event identifier
#' @param expval total amount exposed
#' @param mu mean event loss
#'
#' @return a tidy data.table with Loss, Year and ID. Where a year simulated with zero events will show as "none"
#' @examples 
#' create_ylt(create_elt(eltr::example_elt, ann_rate="rate", mu="mean", 
#'       sdev_i = "sdevi" , sdev_c = "sdevc", expval="exp"),
#'       sims=10,ann_rate = "rate" ,event_id = "id",expval = "exp",mu ="mean")
#' @export

create_ylt <- function(dt, sims, ann_rate, event_id, expval, mu){

  Year <- NULL # avoid the no visible binding for global variable when running R_CMD_CHECK
  set.seed(1)
  yr<-1:sims
  lda <-  sum(dt[, get(ann_rate)])
  sim_events <- stats::rpois(n = sims, lambda = lda)

  row_port <- sample(x = 1:length(dt[,get(event_id)]), size = sum(sim_events), replace = TRUE, prob = dt$random_num)

  event_loss_port<- ifelse( dt$alpha[row_port]>0, stats::rbeta(length(row_port), dt$alpha[row_port], dt$beta[row_port])*dt[,get(expval)][row_port],dt[,get(mu)][row_port])

  lossNo_port <-  rep(yr, times= sim_events)

  event_loss_port <- data.table(EventID = dt[,get(event_id)][row_port], Loss= event_loss_port, Year=lossNo_port)

  zero_port <- yr[!(yr%in% lossNo_port)]

  try(zero_yrs_port <- data.table(Loss=0, Year=zero_port, Event="None"), silent = TRUE)

  EventDT_Port <- data.table(Year=lossNo_port,  Loss=event_loss_port$Loss, Event= dt[,get(event_id)][row_port])
  try(EventDT_Port <- rbind(EventDT_Port, zero_yrs_port), silent=TRUE)


  EventDT_Port <- EventDT_Port[order(Year)]

  EventDT_Port[]


}


#' OEP Curve
#'
#' @param dt aggregate annual YLT
#' @param y vector of year
#' @param z vector of loss amount
#' @param rp return period default points=  c(10000,5000,1000,500,250,200,100,50, 25,10,5 , 2)
#'
#' @return a vector of OEP at return periods as specified by the argument rp
#' @examples
#' create_oep_curve(data.table::data.table("Year" = c(1,2,3,4,5) , 
#'           "Loss" =c(1 , 20 , 500 , 100 , 10000)) , y= "Year", z="Loss")
#' @export

create_oep_curve <- function(dt, y, z, rp =  c(10000,5000,1000,500,250,200,100,50, 25,10,5 , 2)){


  Max_Port <- dt[, max(get(z)) , by=get(y)]

  data.table(return_period=rp, OEP=sapply(1 - 1/rp, function(x) stats::quantile(Max_Port$V1, x)) )

}










